tools/inc/stljobs.h

```diff
@@ -0,0 +1,606 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <assert.h>
+#include <atomic>
+#include <iterator>
+#include <memory>
+#include <stdexcept>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <string>
+#include <string_view>
+#include <userenv.h>
+
+#include <Windows.h>
+
+struct api_exception : std::exception {
+    const char* api;
+    unsigned long lastError;
+
+    explicit api_exception(const char* api_, unsigned long lastError_) noexcept : api(api_), lastError(lastError_){};
+
+    [[noreturn]] void give_up() const {
+        fflush(stdout);
+        fprintf(stderr, "The API %s failed unexpectedly; last error 0x%08lX\n", api, lastError);
+        abort();
+    }
+
+    const char* what() const noexcept override {
+        return "win32 exception";
+    }
+};
+
+[[noreturn]] inline void api_failure(const char* const api, const unsigned long lastError = GetLastError()) {
+    throw api_exception{api, lastError};
+}
+
+struct invalid_handle_value_policy {
+    static constexpr HANDLE Empty = INVALID_HANDLE_VALUE;
+};
+
+struct null_handle_policy {
+    static constexpr HANDLE Empty = 0;
+};
+
+template <class EmptyPolicy>
+class handle {
+public:
+    handle() = default;
+
+    explicit handle(const HANDLE hInitial) noexcept : hImpl(hInitial) {}
+
+    handle(handle&& other) noexcept : hImpl(std::exchange(other.hImpl, EmptyPolicy::Empty)) {}
+
+    handle& operator=(handle&& other) noexcept {
+        handle moved = std::move(other);
+        swap(moved, *this);
+        return *this;
+    }
+
+    handle& operator=(const HANDLE hNew) & noexcept {
+        handle moved{hNew};
+        swap(moved, *this);
+        return *this;
+    }
+
+    ~handle() noexcept {
+        if (hImpl != EmptyPolicy::Empty) {
+            CloseHandle(hImpl);
+        }
+    }
+
+    friend void swap(handle& lhs, handle& rhs) noexcept {
+        using std::swap;
+        swap(lhs.hImpl, rhs.hImpl);
+    }
+
+    void close() noexcept {
+        if (hImpl != EmptyPolicy::Empty) {
+            CloseHandle(hImpl);
+            hImpl = EmptyPolicy::Empty;
+        }
+    }
+
+    [[nodiscard]] explicit operator bool() const noexcept {
+        return hImpl != EmptyPolicy::Empty;
+    }
+
+    [[nodiscard]] HANDLE get() const noexcept {
+        return hImpl;
+    }
+
+    [[nodiscard]] HANDLE detach() noexcept {
+        return std::exchange(hImpl, EmptyPolicy::Empty);
+    }
+
+    [[nodiscard]] HANDLE* out() noexcept {
+        return &hImpl;
+    }
+
+private:
+    HANDLE hImpl{EmptyPolicy::Empty};
+};
+
+inline handle<invalid_handle_value_policy> create_file(const wchar_t* const lpFileName, unsigned long dwDesiredAccess,

```

Comment:
I changed LPCTSTR because that depends on whether `_UNICODE` / `UNICODE` are defined, and I want this to always have `wchar_t` semantics. The other parameters I copied *exactly* as they appear in the documentation for `CreateFile`. (Same for the other similar wrappers like `create_named_pipe` in this file)

Except I guess I translated DWORD... I don't know why I did that.